File: cocos2dAndroidSample/Headless/HeadlessActivity.m
//
//  HeadlessActivity.m
//  Headless
//
//  Created by Philippe Hausler on 6/12/14.
//  Copyright (c) 2014 Apportable. All rights reserved.
//

#import "HeadlessActivity.h"
#import <android/native_window.h>
#import <bridge/runtime.h>

#import "cocos2d.h"

@implementation AndroidViewSurface
@end


@implementation AndroidViewSurfaceHolder
@bridge (instance, method) surface = getSurface;
@end


@implementation HeadlessActivity {
    CCGLView* _glView;
}

@bridge (callback) onCreated: = onCreated;
@bridge (callback) surfaceCreated: = surfaceCreated;
@bridge (callback) surfaceDestroyed: = surfaceDestroyed;
@bridge (callback) surfaceChanged:format:width:height: = surfaceChanged;

- (void)onCreated:(AndroidBundle *)bundle
{
    NSLog(@"Got to here!");
    [[AndroidLooper currentLooper] scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
    [[NSRunLoop currentRunLoop] runUntilDate:[NSDate distantFuture]];
}

- (void)surfaceCreated:(AndroidViewSurfaceHolder*)holder
{
    
}

- (void)surfaceChanged:(AndroidViewSurfaceHolder*)holder format:(int)format width:(int)width height:(int)height
{
    if(_glView != nil)
        return;

    // Note: temp hack, this all needs be abstracted into cocos2d (the same way it works on iOS)
    ANativeWindow* window = ANativeWindow_fromSurface(bridge_getJava(holder.surface));
    
    _glView = [[CCGLView alloc] init];
    [_glView setupView:window];

    CCDirectorAndroid* director = (CCDirectorAndroid*) [CCDirector sharedDirector];
    [director setView:_glView];
    
    
    // set FPS at 60
//	NSTimeInterval animationInterval = [(config[CCSetupAnimationInterval] ?: @(1.0/60.0)) doubleValue];
//    [director setAnimationInterval:animationInterval];
//	director.fixedUpdateInterval = [(config[CCSetupFixedUpdateInterval] ?: @(1.0/60.0)) doubleValue];

    [director setProjection:CCDirectorProjection2D];
    [CCTexture setDefaultAlphaPixelFormat:CCTexturePixelFormat_RGBA8888];
    
    [director reshapeProjection:CGSizeMake(width, height)];
    
    [[CCDirector sharedDirector] startAnimation];
    
}

- (void)surfaceDestroyed:(AndroidViewSurfaceHolder*)holder
{
    
}

@end

File: cocos2dAndroidSample/Headless/src/main/java/com/apportable/HeadlessActivity.java
package com.apportable;

import android.app.Activity;

import android.os.Bundle;
import com.apportable.RuntimeService;
import android.os.Handler;
import android.view.SurfaceView;
import android.view.SurfaceHolder;

import android.util.Log;

//import com.apportable.Headless.R;

public class HeadlessActivity extends Activity implements SurfaceHolder.Callback {

    private RuntimeService mRuntimeService;

    private native void onCreated(Bundle savedInstanceState);
    
    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mRuntimeService = new RuntimeService(this);
        mRuntimeService.loadLibraries();
        
        new Handler().post(new Runnable() {
            @Override
            public void run() {
                SurfaceView surfaceView = new SurfaceView(HeadlessActivity.this);
                surfaceView.getHolder().addCallback(HeadlessActivity.this);
                setContentView(surfaceView);

                onCreated(savedInstanceState);
            }
        });
    }
    
    public native void surfaceCreated(SurfaceHolder holder);
    public native void surfaceChanged(SurfaceHolder holder, int format, int width, int height);
    public native void surfaceDestroyed(SurfaceHolder holder);
}

